ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/RetryingGetBlockFromPeersTask.java

```diff
@@ -14,135 +14,101 @@
   }
 
   public static RetryingGetBlockFromPeersTask create(
-      final ProtocolContext protocolContext,
       final ProtocolSchedule protocolSchedule,
       final EthContext ethContext,
       final MetricsSystem metricsSystem,
       final int maxRetries,
-      final Optional<Hash> hash,
+      final Optional<Hash> maybeHash,
       final long blockNumber) {
     return new RetryingGetBlockFromPeersTask(
-        protocolContext,
-        ethContext,
-        protocolSchedule,
-        metricsSystem,
-        maxRetries,
-        hash,
-        blockNumber);
-  }
-
-  @Override
-  public void assignPeer(final EthPeer peer) {
-    super.assignPeer(peer);
-    triedPeers.add(peer);
+        ethContext, protocolSchedule, metricsSystem, maxRetries, maybeHash, blockNumber);
   }
 
   @Override
-  protected CompletableFuture<AbstractPeerTask.PeerTaskResult<Block>> executePeerTask(
-      final Optional<EthPeer> assignedPeer) {
-
+  protected CompletableFuture<PeerTaskResult<Block>> executeTaskOnCurrentPeer(
+      final EthPeer currentPeer) {
     final GetBlockFromPeerTask getBlockTask =
         GetBlockFromPeerTask.create(
-            protocolSchedule, getEthContext(), blockHash, blockNumber, getMetricsSystem());
-
-    getBlockTask.assignPeer(
-        assignedPeer
-            .filter(unused -> getRetryCount() == 1) // first try with the assigned preferred peer
-            .orElseGet( // then selecting a new one from the pool
-                () -> {
-                  assignPeer(selectNextPeer());
-                  return getAssignedPeer().get();
-                }));
-
-    LOG.debug(
-        "Getting block {} ({}) from peer {}, attempt {}",
-        blockNumber,
-        blockHash,
-        getAssignedPeer(),
-        getRetryCount());
+            protocolSchedule, getEthContext(), maybeBlockHash, blockNumber, getMetricsSystem());
+    getBlockTask.assignPeer(currentPeer);
 
     return executeSubTask(getBlockTask::run)
         .thenApply(
             peerResult -> {
+              debugLambda(
+                  LOG,
+                  "Got block {} from peer {}, attempt {}",
+                  peerResult.getResult()::toLogString,
+                  peerResult.getPeer()::toString,
+                  this::getRetryCount);
               result.complete(peerResult);
               return peerResult;
             });
   }
 
-  private EthPeer selectNextPeer() {
-    return getEthContext()
-        .getEthPeers()
-        .streamBestPeers()
-        .filter(peer -> !triedPeers.contains(peer))
-        .findFirst()
-        .orElseThrow(NoAvailablePeersException::new);
-  }
-
   @Override
   protected boolean isRetryableError(final Throwable error) {
-    return (blockNumber > protocolContext.getBlockchain().getChainHeadBlockNumber())
-        && (super.isRetryableError(error) || error instanceof IncompleteResultsException);
+    return super.isRetryableError(error) || error instanceof IncompleteResultsException;
```

Comment:
do we no longer need to check that the blockNumber is greater than chain head?  or rather should we check that the optional blockHash is already present in the chain?

Reply:
I removed that check because as you said checking the block number alone is not correct, and could prevent downloading legit new blocks with a different hash for the same block number.

About the check in GetBlockFromPeerTask, we usually check before fetching from peers if we have that block, so the chance to re-download a block that we already have is very low, to justify the change.